cmd/test/src/lib.rs

```diff
@@ -227,441 +207,126 @@ impl fmt::Display for TestRunResult {
-                let source = match *port {
-                    0 => TestSource::KernelLog,
-                    1 => TestSource::UserLog,
-                    8 => TestSource::Suite,
-                    _ => {
-                        bail!("spurious data on port {}: {:x?}", port, payload);
-                    }
-                };
-
-                for p in payload {
-                    match testrun.consume(source, *p as char) {
-                        Ok(_) => {}
-                        Err(err) => {
-                            testrun.report(
-                                output,
-                                &wire.borrow(),
-                                Some(&err),
-                            )?;
-                            return Err(err);
-                        }
-                    }
-
-                    if testrun.completed() {
-                        if testrun.failed() {
-                            testrun.report(output, &wire.borrow(), None)?;
-                            std::process::exit(1);
-                        }
-
-                        if subargs.dumpalways {
-                            testrun.report(output, &wire.borrow(), None)?;
-                        }
-
-                        std::process::exit(0);
-                    }
-                }
+        let results = context.run(core, ops.as_slice(), None)?;
+        if results.is_empty() {
+            bail!("Bad return");
+        }
 
-                Ok(())
-            }
-            ITMPayload::None => {
-                match packet.header {
-                    ITMHeader::Sync => {
-                        if !kicked {
-                            shared.borrow_mut().halt()?;
-                            shared.borrow_mut().write_word_32(v.addr, 1)?;
-                            shared.borrow_mut().run()?;
-                            kicked = true;
-                        }
-                    }
-                    ITMHeader::Malformed(datum) => {
-                        bail!("malformed datum: 0x{:x}", datum);
-                    }
-                    _ => {}
+        let result = match &results[0] {
+            Ok(s) => {
+                if s[0] == 1 {
+                    TestResult::Ok
+                } else {
+                    TestResult::Fail
                 }
-
-                Ok(())
             }
-            _ => {
-                bail!("unknown packet: {:x?}", packet);
-            }
-        },
-    );
+            Err(e) => TestResult::Unknown(format!("{}", e)),
+        };
 
-    match rval {
-        Ok(_) => rval,
-        Err(err) => {
-            testrun.report(output, &wire.borrow(), Some(&err))?;
-            Err(err)
-        }
+        println!("{:#}", result);
+        writeln!(
+            out,
+            "==== Test {} result: {:?}",
+            test_name,
+            result.to_str_no_color()
+        )?;
+        writeln!(out, "==== Task state")?;
+
+        cmd_tasks::print_tasks(
```

Comment:
I think that this is fine for now, but we may want to stick this into a different crate if we find we're using it a bunch....